core/domain/story_services.py

```diff
@@ -617,3 +621,258 @@ def record_completed_node_in_story_context(user_id, story_id, node_id):
     if node_id not in progress_model.completed_node_ids:
         progress_model.completed_node_ids.append(node_id)
         progress_model.put()
+
+
+def get_story_rights_from_model(story_rights_model):
+    """Constructs a StoryRights object from the given story rights model.
+
+    Args:
+        story_rights_model: StoryRightsModel. Story rights from the
+            datastore.
+
+    Returns:
+        StoryRights. The rights object created from the model.
+    """
+
+    return story_domain.StoryRights(
+        story_rights_model.id,
+        story_rights_model.manager_ids,
+        story_rights_model.story_is_published
+    )
+
+
+def publish_story(story_id, committer_id):
+    """Marks the given story as published.
+
+    Args:
+        story_id: str. The id of the given story.
+        committer_id: str. ID of the committer.
+
+    Raises:
+        Exception. The given story does not exist.
+        Exception. The story is already published.
+        Exception. The user does not have enough rights to publish the story.
+    """
+    story = get_story_by_id(story_id, strict=False)
+    for node in story.story_contents.nodes:
+        if not node.exploration_id:
+            raise Exception('Story node does not contain an exploration id.')
+        if not exp_services.get_exploration_by_id(
+                node.exploration_id, strict=False):
+            raise Exception(
+                'Story node with exploration id %s doesn\'t exist.'
+                % node.exploration_id)
+        exploration_rights = rights_manager.get_exploration_rights(
+            node.exploration_id, strict=False)
+        if exploration_rights.is_private():
+            raise Exception(
+                'Story node with exploration id %s isn\'t published.'
```

Comment:
In the case that an exploration gets unpublished after the story is published, what happens?

If we want to allow the story to remain published even after an exploration in it is unpublished, I wonder if we should just allow stories to be published despite containing unpublished explorations. I think the only real requirement could be that at least one exploration is published. The reason for this is that we may want to further develop published stories, so the story author can continue working on expanding the story outline while learners play existing nodes in the story.

Reply:
In this case, currently, if the story has all the valid nodes with published exploration ids then only it'll be published. I suppose you're correct about the approach. We should allow publishing of stories with published or unpublished exploration alike. While presenting the story to the learner it could be taken care of what node should be shown to the learner.

Reply:
Considering the above discussion, I don't think that we need any of the checks then? As the story may still be under construction when it is being published.

Reply:
/cc @aks681

Reply:
But what if the explorations corresponding to initial story nodes are unpublished and those corresponding to later ones are published. In this case, surfacing the story to a learner wouldn't make sense, right? As effectively the learner cannot start the story. I think there should be a check that validates whether the initial story node is published or not. If at least  that is published, then we can surface the story.  What do you think?

Reply:
Sounds good. I've updated it accordingly. Thanks!